domain/status/service/storage.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	corestatus "github.com/juju/juju/core/status"
+	"github.com/juju/juju/core/storage"
+	"github.com/juju/juju/core/trace"
+	"github.com/juju/juju/domain/status"
+	"github.com/juju/juju/internal/errors"
+)
+
+// StorageState provides access to storage related state methods.
+type StorageState interface {
+	// SetFilesystemStatus saves the given filesystem status, overwriting any
+	// current status data. The following errors can be expected:
+	// - [storageerrors.FilesystemNotFound] if the filesystem doesn't exist.
+	SetFilesystemStatus(
+		ctx context.Context,
+		filesystemUUID storage.FilesystemUUID,
+		sts status.StatusInfo[status.StorageFilesystemStatusType],
+	) error
+
+	// GetFilesystemUUIDByID returns the filesystem UUID for the given
+	// filesystem unique id. If no filesystem is found, an error satisfying
+	// [storageerrors.FilesystemNotFound] is returned.
+	GetFilesystemUUIDByID(ctx context.Context, id string) (storage.FilesystemUUID, error)
+
+	// SetVolumeStatus saves the given volume status, overwriting any
+	// current status data. The following errors can be expected:
+	// - [storageerrors.VolumeNotFound] if the volume doesn't exist.
+	SetVolumeStatus(
+		ctx context.Context,
+		volumeUUID storage.VolumeUUID,
+		sts status.StatusInfo[status.StorageVolumeStatusType],
+	) error
+
+	// GetVolumeUUIDByID returns the volume UUID for the given
+	// volume unique id. If no volume is found, an error satisfying
+	// [storageerrors.VolumeNotFound] is returned.
+	GetVolumeUUIDByID(ctx context.Context, id string) (storage.VolumeUUID, error)
+}
+
+// SetFilesystemStatus saves the given filesystem status, overwriting any
+// current status data. If returns an error satisfying
+// [storageerrors.FilesystemNotFound] if the filesystem doesn't exist.
+func (s *Service) SetFilesystemStatus(
```

Comment:
Both filesystem and volume use a state method that checks if the transition is to error and then checks that a message has been set. This message check must be lifted up to the service layer as this is not a transition validation check and is the responsibility of service layer.

We're in this place because of copy and pasting from Mongo state.

Reply:
The error message check can be uplifted. The check involving pending needs to stay in state. Originally the checks were kept together so they were all in one place. But they are now split.